core/src/main/scala/io/odin/loggers/ConditionalLogger.scala

```diff
@@ -0,0 +1,80 @@
+package io.odin.loggers
+
+import cats.MonadError
+import cats.effect.{Concurrent, ContextShift, ExitCase, Resource, Timer}
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.order._
+import io.odin.{Level, Logger, LoggerMessage}
+import monix.catnap.ConcurrentQueue
+import monix.execution.{BufferCapacity, ChannelType}
+
+final case class ConditionalLogger[F[_]: Timer] private (
+    queue: ConcurrentQueue[F, LoggerMessage],
+    inner: Logger[F],
+    override val minLevel: Level
+)(implicit F: MonadError[F, Throwable])
+    extends DefaultLogger[F](minLevel) {
+
+  def log(msg: LoggerMessage): F[Unit] =
+    queue.tryOffer(msg).void
+
+  private def drain(exitCase: ExitCase[Throwable]): F[Unit] = {
+    val level = exitCase match {
+      case ExitCase.Completed => inner.minLevel
+      case _                  => minLevel
+    }
+
+    queue
+      .drain(0, Int.MaxValue)
+      .flatMap(msgs => inner.log(msgs.filter(_.level >= level).toList))
+      .attempt
+      .void
+  }
+
+}
+
+object ConditionalLogger {
+
+  /**
+    * Create ConditionalLogger that buffers messages and sends them to the inner logger when the resource is released.
+    * If evaluation of the bracket completed with an error, the `fallbackLevel` is used as a `minLevel`.
+    *
+    * Example:
+    * {{{
+    *   consoleLogger[F](minLevel = Level.Info)
+    *     .conditional(Level.Debug)
+    *     .use(logger => logger.debug("debug message") >> trickyCode)
+    * }}}
+    *
+    * If evaluation completed with an error, the messages with `level >= Level.Debug` will be sent to an inner logger.
+    * If evaluation completed successfully, the messages with `level >= Level.Info` will be sent to an inner logger.
+    *
+    *
+    * '''Important:''' nothing is logged until the resource is released.
+    * Example:
+    * {{{
+    * consoleLogger[F](minLevel = Level.Info)
+    *   .conditional(Level.Debug)
+    *   .use(logger => logger.info("conditional logger") >> Timer[F].sleep(10.seconds))
+    * }}}
+    *
+    * The message will be logged after 10 seconds. Thus use the logger with caution.
+    *
+    * @param inner logger that will receive messages from the buffer
+    * @param fallbackLevel min level that will be used in case of error
+    */
+  def create[F[_]: Timer: Concurrent: ContextShift](inner: Logger[F], fallbackLevel: Level): Resource[F, Logger[F]] = {
+    def acquire: F[ConditionalLogger[F]] =
+      for {
+        queue <- ConcurrentQueue.withConfig[F, LoggerMessage](BufferCapacity.Unbounded(), ChannelType.MPSC)
```

Comment:
I feel we should provide a way to optionally limit queue size for some cases when users would produce tons of debug logs and unknowingly consume all the memory.

Reply:
Agreed